Add new Waffle tutorial (#1642)
ethereum/ethereum-org-website: src/content/developers/tutorials/testing-smart-contract-with-waffle/index.md

---

> title: Testing smart contract with Waffle

Reviewer: ```suggestion
title: Testing smart contracts with Waffle
```

> description: Tutorial with examples of using smart contract matchers with waffle

Reviewer: ```suggestion
description: Learn how test Solidity smart contracts and use smart contract matchers with Waffle.
```

author: Vladislav Starostenko
tags: ["waffle", "smart contracts", "solidity", "testing"]
skill: intermediate
lang: en
sidebar: true
published: 2020-10-16
---

In this tutorial you will learn how to:

- Write tests for smart contracts with Waffle
- Use some popular matchers to test smart contracts with Waffle

Assumptions:

- you can get around in a terminal,
- you can create a new `JavaScript` project,
- you've written a few lines of `Solidity` code,
- you've written a few tests in `JavaScript`,
- you’ve used `yarn` or `npm`, JavaScripts’s package installer.

Again, if any of these are untrue, or you don’t plan to reproduce the code in this article, you can likely still follow along just fine.

## A few words about Waffle {#a-few-words-about-waffle}

[Waffle](https://getwaffle.io) is the most advanced library for writing and testing smart contracts.

> Works with ethers-js.

Reviewer: ```suggestion
Works with the [JavaScript API](/developers/docs/apis/javascript/) ethers-js.
```

> More details you can read in [waffle documentation](https://ethereum-waffle.readthedocs.io/en/latest/#waffle-documentation) !

Reviewer: ```suggestion
You can read more details in the [Waffle documentation](https://ethereum-waffle.readthedocs.io/en/latest/#waffle-documentation) !
```

## The quick tutorial {#the-quick-tutorial}

First things first, create new `JavaScript` or `TypeScript` project ( I'll use `TS`, but if you use `JS` it's not a problem ) :

Somewhat like this :

<details>

> <summary>package.json</summary>

Reviewer: Didn't know about `<details>` & `<summary>` elements - neat!

        {
          "requireStringLiterals": true
        }
      ],
        "wrap-iife": [
        "error",
        "any",
        {
          "functionPrototypeMethods": true
        }
      ],
        "yield-star-spacing": [
        "error",
        "both"
      ],
        "yoda": [
        "error",
        "never"
      ]
    },
      "overrides": [
      {
        "files": [
          "test/**/*.ts"
        ],
        "rules": {
          "@typescript-eslint/no-explicit-any": "off",
          "@typescript-eslint/no-non-null-assertion": "off",
          "@typescript-eslint/no-var-requires": "off",
          "no-unused-expressions": "off",
          "prefer-promise-reject-errors": "off",
          "import/no-extraneous-dependencies": [
            "error",
            {
              "devDependencies": true
            }
          ]
        }
      }
    ]
    }

</details>

## Step #1: Install waffle in your project [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#installation) {#step-1-install-waffle-in-your-project}

To get started, install `ethereum-waffle`. In this tutorial, I'll use `yarn`, so to install `ethereum-waffle` run:

> ```

Reviewer: ```suggestion
```bash
```

 yarn add --dev ethereum-waffle
```

## Step #2: Write a smart contract [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#writing-a-contract) {#step-2-write-a-smart-contract}

In this tutorial, I'll use [ERC20](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/ded2b0a55c9c13731963ab7b85a70c8e73504bab/contracts/token/ERC20/ERC20.sol) token from [OpenZeppelin](https://openzeppelin.com).

So, add `OpenZeppelin` by installing it with `yarn`:

> ```

Reviewer: ```suggestion
```bash
```

 yarn add @openzeppelin/contracts -D
```

Then create `BasicToken.sol` contract in `src` directory:

```solidity
pragma solidity ^0.6.0;

import "@openzeppelin/contracts/token/ERC20/ERC20.sol";

// Example class - a mock class using delivering from ERC20
contract BasicToken is ERC20 {
    constructor(uint256 initialBalance) ERC20("Basic", "BSC") public {
        _mint(msg.sender, initialBalance);
    }
}

```

## Step #3: Compile your smart contract [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#compiling-the-contract) {#step-3-compile-your-smart-contract}

To compile your smart contract add the following entry in the `package.json` of your project :

```json
{
  "scripts": {
    "build": "waffle"
  }
}
```

Also, add `waffle.json` file in the main directory of your project.

An example of `waffle.json` configuration:

```json
{
  "compilerType": "solcjs",
  "compilerVersion": "0.6.2",
  "sourceDirectory": "./src",
  "outputDirectory": "./build"
}
```

> You can read more about the waffle configuration [here](https://ethereum-waffle.readthedocs.io/en/latest/configuration.html#configuration).

Reviewer: ```suggestion
You can read more about the Waffle configuration [here](https://ethereum-waffle.readthedocs.io/en/latest/configuration.html#configuration).
```

    }

</details>

## Step #4: Test your smart contract [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#writing-tests) {#step-4-test-your-smart-contract}

### Step #4.1 Install necessary dependencies [Link to doc](https://ethereum-waffle.readthedocs.io/en/latest/getting-started.html#writing-tests)

After we have successfully authored a Smart Contract we can test it. We will use `Waffle` to do it.

Tests in `Waffle` are written using `Mocha` alongside with `Chai`. We can use a different test environment, but `Waffle` matchers only work with `Chai`.

So, we need to add `Chai` to our dependencies :

> ```

Reviewer: ```suggestion
```bash
```
